Readme.md

```diff
@@ -41,6 +41,9 @@ There a three types of constructors available:
 - `new Fraction (<numerator>, <denominator>)` using `BigInteger` for numerator and denominator.
 - `new Fraction (<numerator>, <denominator>, <reduce>)` using `BigInteger` for numerator and denominator + `bool` to indicate if the resulting fraction shall be normalized (reduced).
 
+> [!IMPORTANT]
+> When creating improper fractions (by specifying the parameter `reduce: false`), please be sure to refer to the [Working with non-normalized fractions](#working-with-non-normalized-fractions) and the [Equality operators](#equality-operators) section for more information about the (side) effects.

```

Comment:
I have no particular preference, but have you considered actually making the Equality/HashCode based on reduced terms? Given that `123.00m` is considered equal to `123m` I don't think it would be inconsistent to say that `12300/100` is equal to `123/1` .

Reply:
I cannot estimate the performance impact on existing code. Currently, most users are likely to use normalized fractions anyway, then the following applies:

If `State == FractionState.IsNormalized` or `_normalizationNotApplied == false` is set for both fractions, then the numerator and denominator can simply be compared. 

However, the situation is different with non-normalized fractions. The *expensive* method `Reduce` would have to be called every time for both `GetHashCode` and `Equals`. And we cannot cache the results (readonly struct).

Another question would be: What if for a use case the following applies: $\frac{1}{2} \neq \frac{2}{4}$ (because these are different representations in the user interface)?

Reply:
I actually had a long discussion with my colleagues back then. The majority believed that `Equals` should always be exact. Microsoft does not follow this rule. In our software we have a problem with this: 
![image](https://github.com/danm-de/Fractions/assets/3825484/317351c4-ca74-409b-a91a-c07519e6c121)

Reply:
If the Equals is `true` then the HashCode should be equal as well, the opposite is not a given.

> I cannot estimate the performance impact on existing code. Currently, most users are likely to use normalized fractions anyway, then the following applies:

The GetHashCode is mostly used when checking for stuff like `collection.Contains(x)` (where collection is actually hash-based)- otherwise we hit the Equals check, which for our scenario can be quite efficient. For actual GetHashCode- related checks, well as you said- it's more likely that somebody is doing the something custom, and can override the way the hash-code is calculated. 
The unintentional / unexpected non-equality is what worries me a little- but we'll be fine :)

Reply:
I added the `.Dump()` output in no particular order, but of course: `Equals(a,b) == true` implied `GetHashCode(a) == GetHashCode(b)`. In the example we have two different time zones - why are both values ​​the same? If I want to know if it is the same *point in time*, I would apply `.ToUniversalTime()` to both variables and then compare.

But hey, I'm open to everything. You said you wanted to use the data type in another project (https://github.com/angularsen/UnitsNet right?). What is the opinion there? Is my imagined “use case” perhaps nonsense?

Reply:
Oh, we've had lots of issues with Equality contract and how to make it so that "1 kg" == "1000 g". My personal feeling is that _making it custom_ is way easier then _making it work for everybody_.. I asked the AI some time ago about a design pattern to justify an argument- it cited the _principle of least surprise_ which I quite liked. So as soon as we open the door the _optimized operations_ someone (new) might get _surprised_.

Reply:
POLS, I remember it. Geoffrey James has been quoted quite a bit.

Reply:
Let me sleep and think about it one night.

Reply:
Started: https://github.com/danm-de/Fractions/tree/feature/default-to-value-equality

Having fun because of NaN != NaN 🥲 

![image](https://github.com/danm-de/Fractions/assets/3825484/ff7d93e1-0756-45cf-b23c-616cc4b7c5d0)

Reply:
Yeah, I don't know whoever thought it was necessary to have the Equals method differ from the == on this one 🤔